core/iwasm/fast-jit/fe/jit_emit_variable.c

```diff
@@ -161,15 +161,150 @@ jit_compile_op_tee_local(JitCompContext *cc, uint32 local_idx)
     return false;
 }
 
+static uint8
+get_global_type(const WASMModule *module, uint32 global_idx)
+{
+    return module->globals[global_idx].type;
+}
+
+static JitReg
+get_data_offset_reg(JitCompContext *cc, uint32 global_idx)
+{
+    JitReg globals, global_inst, data_offset;
+
+    globals = get_globals_reg(cc->jit_frame);
+
+    /* global_inst = globals + global_idx */
+    global_inst = jit_cc_new_reg_ptr(cc);
+    GEN_INSN(ADD, global_inst, globals,
+             NEW_CONST(I64, global_idx * sizeof(WASMGlobalInstance)));
+
+    /* data_offset = global_inst->data_offset */
+    data_offset = jit_cc_new_reg_I64(cc);
+    GEN_INSN(LDI32, data_offset, global_inst,
+             NEW_CONST(I32, offsetof(WASMGlobalInstance, data_offset)));
+
+    return data_offset;
+}
+
 bool
 jit_compile_op_get_global(JitCompContext *cc, uint32 global_idx)
 {
+    uint8 global_type = 0;
+    JitReg value = 0;
+
+    if (global_idx >= cc->cur_wasm_module->global_count) {
+        jit_set_last_error(cc, "the global index is out of range");
+        goto fail;
+    }
+
+    global_type = get_global_type(cc->cur_wasm_module, global_idx);
+    switch (global_type) {
+        case VALUE_TYPE_I32:
+#if WASM_ENABLE_REF_TYPES != 0
+        case VALUE_TYPE_EXTERNREF:
+        case VALUE_TYPE_FUNCREF:
+#endif
+        {
+            value = jit_cc_new_reg_I32(cc);
+            GEN_INSN(LDI32, value, get_global_data_reg(cc->jit_frame),
+                     get_data_offset_reg(cc, global_idx));
```

Comment:
No need to load global data offset from global instance, it can be gotten from cc->cur_wasm_module in compilation time:
```C
uint32 global_data_offset;

if (global_idx < cur_wasm_module->import_global_count) {
    WASMImport *import = module->import_globals + global_idx;
    WASMGlobalImport *global_import = &import->u.global;
    global_data_offset = global_import->data_offset;
}
else {
    WASMGlobal *global = cur_wasm_module->globals +
                     + (global_idx - cur_wasm_module->import_global_count);
    global_data_offset = global->data_offset;
}
```
Then here:
```
            GEN_INSN(LDI32, value, get_global_data_reg(cc->jit_frame),
                     NEW_CONST(I32, global_data_offset));
```

Reply:
not sure about that. It seems we don't know about `global_data_offset` before instantiation.

```c
typedef struct WASMGlobalImport {
    char *module_name;
    char *field_name;
    uint8 type;
    bool is_mutable;
    /* global data after linked */
    WASMValue global_data_linked;
    bool is_linked;
#if WASM_ENABLE_MULTI_MODULE != 0
    /* imported function pointer after linked */
    /* TODO: remove if not needed */
    WASMModule *import_module;
    WASMGlobal *import_global_linked;
#endif
} WASMGlobalImport;

struct WASMGlobal {
    uint8 type;
    bool is_mutable;
    InitializerExpression init_expr;
};
```

Reply:
Sorry for that, now I know why you get global data offset from module inst, we haven't calculated it in loader stage. But definitely we can do that, in fact in the AOT compilation, we pre-calculate the data offset of each global when converting wasm module into AOT compilation data, and directly use it in AOT compilation:

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot.c#L181-L182

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot.c#L215-L216

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot_emit_variable.c#L124-L134

Can we add field `uint32 data_offset` for WASMGlobalImport and WASMGlobal, and calculate them in wasm_loader.c (or before fast jit compilation) so as to improve the performance?